plugins/spectate.cpp

```diff
@@ -66,23 +145,77 @@ DFhackCExport command_result plugin_enable(color_ostream &out, bool enable) {
         EM::registerListener(EventType::JOB_STARTED, start, plugin_self);
         EM::registerListener(EventType::JOB_COMPLETED, complete, plugin_self);
     } else if (!enable && enabled) {
+        // warp 8, engage!
         out.print("Spectate mode disabled!\n");
         EM::unregisterAll(plugin_self);
         job_tracker.clear();
         freq.clear();
     }
     enabled = enable;
-    return CR_OK;
+    return DFHack::CR_OK;
 }
 
 command_result spectate (color_ostream &out, std::vector <std::string> & parameters) {
-    return plugin_enable(out, !enabled);
+    // todo: parse parameters
+    if(!parameters.empty()) {
+        if (parameters[0] == "disable") {
+            return plugin_enable(out, false);
+        } else if (parameters[0] == "enable") {
+            return plugin_enable(out, true);
+        } else if (parameters[0] == "godmode") {
+            out.print("todo?\n"); // todo: adventure as deity?
+        } else if (parameters[0] == "auto-unpause") {
+            dismiss_pause_events = !dismiss_pause_events;
+
+            // update the announcement settings if we can
+            if (dismiss_pause_events) {
+                if (World::SaveAnnouncementSettings()) {
+                    World::DisableAnnouncementPausing();
+                    pause_lock->lock();
+                } else {
+                    lock_collision = true;
+                }
+            } else {
+                pause_lock->unlock();
+                if (!World::RestoreAnnouncementSettings()) {
+                    // this in theory shouldn't happen, if others use the lock like we do in spectate
+                    lock_collision = true;
+                }
+            }
+
+            // report to the user how things went
+            if (!lock_collision){
+                out.print(dismiss_pause_events ? "auto-unpause: on\n" : "auto-unpause: off\n");
+            } else {
+                out.print("auto-unpause: must wait for another Pausing::AnnouncementLock to be lifted.\n");
+            }
+
+            // probably a typo
+            if (parameters.size() == 2) {
+                out.print("If you want additional options open an issue on github, or mention it on discord.\n");
+                return DFHack::CR_WRONG_USAGE;
+            }
+        }
+    } else {
+        out.print(enabled ? "Spectate is enabled.\n" : "Spectate is disabled.\n");
+        if(enabled) {
+            out.print(dismiss_pause_events ? "auto-unpause: on.\n" : "auto-unpause: off.\n");
+        }
+    }
+    return DFHack::CR_OK;
 }
 
+// every tick check whether to decide to follow a dwarf
 void onTick(color_ostream& out, void* ptr) {
+    if (!df::global::ui) return;
```

Comment:
Not strictly necessary with REQUIRE_GLOBAL() - the plugin simply won't load if `ui` is null

Reply:
Oh. Neat